From 565dbbaef48df9d5f24c526219a52acf06753e77 Mon Sep 17 00:00:00 2001
From: Guoniu Zhou <guoniu.zhou@nxp.com>
Date: Mon, 28 Apr 2025 10:41:40 +0800
Subject: [PATCH] LF-14772: media: imx8-isi: Add 4K support when bypass ISI
 channel

According to the info from IC team, 4K won't need ISI adjacent
line buffer if bypass ISI channel. But in current driver version,
it will always need adjacent line buffer when input size is more
than 2K, so remove this limitation to align with ISI feature.

Signed-off-by: Guoniu Zhou <guoniu.zhou@nxp.com>
Reviewed-by: Robby Cai <robby.cai@nxp.com>
---
.../media/platform/nxp/imx8-isi/imx8-isi-core.h    |   1 +
drivers/media/platform/nxp/imx8-isi/imx8-isi-m2m.c |   2 +-
.../media/platform/nxp/imx8-isi/imx8-isi-pipe.c    | 106 +++++++++++++++------
.../media/platform/nxp/imx8-isi/imx8-isi-video.c   |   5 +-
4 files changed, 81 insertions(+), 33 deletions(-)


diff --git a/drivers/media/platform/nxp/imx8-isi/imx8-isi-core.h b/drivers/media/platform/nxp/imx8-isi/imx8-isi-core.h
index fcf52ba0d015..007269aedbf1 100644
--- a/drivers/media/platform/nxp/imx8-isi/imx8-isi-core.h
+++ b/drivers/media/platform/nxp/imx8-isi/imx8-isi-core.h
@@ -262,6 +262,7 @@ struct mxc_isi_pipe {
 	u8				acquired_res;
 	u8				chained_res;
 	bool				chained;
+	bool				bypass;
 
 	/* Virtual channel ID for the ISI channel */
 	u8				vc;
diff --git a/drivers/media/platform/nxp/imx8-isi/imx8-isi-m2m.c b/drivers/media/platform/nxp/imx8-isi/imx8-isi-m2m.c
index d9885c449b6f..f5d289ba66fb 100644
--- a/drivers/media/platform/nxp/imx8-isi/imx8-isi-m2m.c
+++ b/drivers/media/platform/nxp/imx8-isi/imx8-isi-m2m.c
@@ -531,7 +531,7 @@ static int mxc_isi_m2m_streamon(struct file *file, void *fh,
 	 * Allocate resources for the channel, counting how many users require
 	 * buffer chaining.
 	 */
-	if (!ctx->chained && out_pix->width > MXC_ISI_MAX_WIDTH_UNCHAINED) {
+	if (!ctx->chained && out_pix->width > MXC_ISI_MAX_WIDTH_UNCHAINED && !bypass) {
 		ret = mxc_isi_channel_chain(m2m->pipe, bypass);
 		if (ret)
 			goto deinit;
diff --git a/drivers/media/platform/nxp/imx8-isi/imx8-isi-pipe.c b/drivers/media/platform/nxp/imx8-isi/imx8-isi-pipe.c
index f1dac906bdc9..92df8598b743 100644
--- a/drivers/media/platform/nxp/imx8-isi/imx8-isi-pipe.c
+++ b/drivers/media/platform/nxp/imx8-isi/imx8-isi-pipe.c
@@ -519,6 +519,62 @@ static int mxc_isi_pipe_enum_mbus_code(struct v4l2_subdev *sd,
 	return -EINVAL;
 }
 
+static inline bool is_chan_bypass(const struct v4l2_mbus_framefmt *sink_fmt,
+				  const struct v4l2_mbus_framefmt *src_fmt,
+				  const struct mxc_isi_bus_format_info *sink_info,
+				  const struct mxc_isi_bus_format_info *src_info)
+{
+	return sink_fmt->width == src_fmt->width &&
+	       sink_fmt->height == src_fmt->height &&
+	       sink_info->encoding == src_info->encoding;
+
+}
+
+static void update_pads_format(struct mxc_isi_pipe *pipe,
+			       struct v4l2_subdev_state *state,
+			       struct v4l2_subdev_format *fmt)
+{
+	struct v4l2_mbus_framefmt *mf = &fmt->format;
+	const struct mxc_isi_bus_format_info *sink_info;
+	const struct mxc_isi_bus_format_info *src_info;
+	struct v4l2_mbus_framefmt *sink_fmt;
+	struct v4l2_mbus_framefmt *src_fmt;
+	struct v4l2_rect *rect;
+	unsigned int max_width;
+	bool bypass;
+
+	sink_fmt = mxc_isi_pipe_get_pad_format(pipe, state, MXC_ISI_PIPE_PAD_SINK);
+	src_fmt = mxc_isi_pipe_get_pad_format(pipe, state, MXC_ISI_PIPE_PAD_SOURCE);
+
+	sink_info = mxc_isi_bus_format_by_code(sink_fmt->code, MXC_ISI_PIPE_PAD_SINK);
+	src_info = mxc_isi_bus_format_by_code(src_fmt->code, MXC_ISI_PIPE_PAD_SOURCE);
+
+	bypass = is_chan_bypass(sink_fmt, src_fmt, sink_info, src_info);
+	pipe->bypass = bypass;
+
+	/*
+	 * Limit the max line length if there's no adjacent pipe to
+	 * chain with and not bypass ISI channel.
+	 */
+	max_width = (!bypass && pipe->id == pipe->isi->pdata->num_channels - 1)
+		  ? MXC_ISI_MAX_WIDTH_UNCHAINED
+		  : MXC_ISI_MAX_WIDTH_CHAINED;
+
+	sink_fmt->width = clamp(sink_fmt->width, MXC_ISI_MIN_WIDTH, max_width);
+
+	/*
+	 * Need to update pad format since its value may out of bound.
+	 */
+	rect = mxc_isi_pipe_get_pad_compose(pipe, state, MXC_ISI_PIPE_PAD_SINK);
+	rect->width = sink_fmt->width;
+
+	src_fmt->width = clamp(src_fmt->width, MXC_ISI_MIN_WIDTH, max_width);
+	rect = mxc_isi_pipe_get_pad_crop(pipe, state, MXC_ISI_PIPE_PAD_SOURCE);
+	rect->width = src_fmt->width;
+
+	*mf = *mxc_isi_pipe_get_pad_format(pipe, state, fmt->pad);
+}
+
 static int mxc_isi_pipe_set_fmt(struct v4l2_subdev *sd,
 				struct v4l2_subdev_state *state,
 				struct v4l2_subdev_format *fmt)
@@ -533,8 +589,6 @@ static int mxc_isi_pipe_set_fmt(struct v4l2_subdev *sd,
 		return -EBUSY;
 
 	if (fmt->pad == MXC_ISI_PIPE_PAD_SINK) {
-		unsigned int max_width;
-
 		info = mxc_isi_bus_format_by_code(mf->code,
 						  MXC_ISI_PIPE_PAD_SINK);
 		if (!info)
@@ -542,17 +596,18 @@ static int mxc_isi_pipe_set_fmt(struct v4l2_subdev *sd,
 							  MXC_ISI_PIPE_PAD_SINK);
 
 		/*
-		 * Limit the max line length if there's no adjacent pipe to
-		 * chain with.
-		 */
-		max_width = pipe->id == pipe->isi->pdata->num_channels - 1
-			  ? MXC_ISI_MAX_WIDTH_UNCHAINED
-			  : MXC_ISI_MAX_WIDTH_CHAINED;
-
-		mf->code = info->mbus_code;
-		mf->width = clamp(mf->width, MXC_ISI_MIN_WIDTH, max_width);
-		mf->height = clamp(mf->height, MXC_ISI_MIN_HEIGHT,
-				   MXC_ISI_MAX_HEIGHT);
+		 * Source and sink pad media bus format are always equal
+		 * when user set sink pad format since the sink pad bus
+		 * format will be propagated to the source pad, so the
+		 * maximum width for ISI channel is 4K. No adjacent line
+		 * buffer is needed.
+ 		 */
+
+ 		mf->code = info->mbus_code;
+		mf->width = clamp(mf->width, MXC_ISI_MIN_WIDTH,
+				  MXC_ISI_MAX_WIDTH_CHAINED);
+ 		mf->height = clamp(mf->height, MXC_ISI_MIN_HEIGHT,
+ 				   MXC_ISI_MAX_HEIGHT);
 		mf->field = V4L2_FIELD_NONE;
 
 		/* Propagate the format to the source pad. */
@@ -609,6 +664,8 @@ static int mxc_isi_pipe_set_fmt(struct v4l2_subdev *sd,
 	format = mxc_isi_pipe_get_pad_format(pipe, state, fmt->pad);
 	*format = *mf;
 
+	update_pads_format(pipe, state, fmt);
+
 	dev_dbg(pipe->isi->dev, "pad%u: code: 0x%04x, %ux%u",
 		fmt->pad, mf->code, mf->width, mf->height);
 
@@ -859,6 +916,9 @@ int mxc_isi_pipe_init(struct mxc_isi_dev *isi, unsigned int id)
 	if (ret < 0)
 		goto error;
 
+	/* ISI channel is bypassed by default */
+	pipe->bypass = true;
+
 	/* Register IRQ handler. */
 	mxc_isi_channel_irq_clear(pipe);
 
@@ -895,36 +955,22 @@ void mxc_isi_pipe_cleanup(struct mxc_isi_pipe *pipe)
 int mxc_isi_pipe_acquire(struct mxc_isi_pipe *pipe,
 			 mxc_isi_pipe_irq_t irq_handler)
 {
-	const struct mxc_isi_bus_format_info *sink_info;
-	const struct mxc_isi_bus_format_info *src_info;
 	struct v4l2_mbus_framefmt *sink_fmt;
-	const struct v4l2_mbus_framefmt *src_fmt;
 	struct v4l2_subdev *sd = &pipe->sd;
 	struct v4l2_subdev_state *state;
-	bool bypass;
 	int ret;
 
 	state = v4l2_subdev_lock_and_get_active_state(sd);
 	sink_fmt = v4l2_subdev_get_try_format(sd, state, MXC_ISI_PIPE_PAD_SINK);
-	src_fmt = v4l2_subdev_get_try_format(sd, state, MXC_ISI_PIPE_PAD_SOURCE);
 	v4l2_subdev_unlock_state(state);
 
-	sink_info = mxc_isi_bus_format_by_code(sink_fmt->code,
-					       MXC_ISI_PIPE_PAD_SINK);
-	src_info = mxc_isi_bus_format_by_code(src_fmt->code,
-					      MXC_ISI_PIPE_PAD_SOURCE);
-
-	bypass = sink_fmt->width == src_fmt->width &&
-		 sink_fmt->height == src_fmt->height &&
-		 sink_info->encoding == src_info->encoding;
-
-	ret = mxc_isi_channel_acquire(pipe, irq_handler, bypass);
+	ret = mxc_isi_channel_acquire(pipe, irq_handler, pipe->bypass);
 	if (ret)
 		return ret;
 
 	/* Chain the channel if needed for wide resolutions. */
-	if (sink_fmt->width > MXC_ISI_MAX_WIDTH_UNCHAINED) {
-		ret = mxc_isi_channel_chain(pipe, bypass);
+	if (sink_fmt->width > MXC_ISI_MAX_WIDTH_UNCHAINED && !pipe->bypass) {
+		ret = mxc_isi_channel_chain(pipe, pipe->bypass);
 		if (ret)
 			mxc_isi_channel_release(pipe);
 	}
diff --git a/drivers/media/platform/nxp/imx8-isi/imx8-isi-video.c b/drivers/media/platform/nxp/imx8-isi/imx8-isi-video.c
index 21e97a652cd0..437d77bf29d0 100644
--- a/drivers/media/platform/nxp/imx8-isi/imx8-isi-video.c
+++ b/drivers/media/platform/nxp/imx8-isi/imx8-isi-video.c
@@ -494,7 +494,7 @@ mxc_isi_format_try(struct mxc_isi_pipe *pipe, struct v4l2_pix_format_mplane *pix
 	unsigned int max_width;
 	unsigned int i;
 
-	max_width = pipe->id == pipe->isi->pdata->num_channels - 1
+	max_width = (!pipe->bypass && pipe->id == pipe->isi->pdata->num_channels - 1)
 		  ? MXC_ISI_MAX_WIDTH_UNCHAINED
 		  : MXC_ISI_MAX_WIDTH_CHAINED;
 
@@ -1327,6 +1327,7 @@ static int mxc_isi_video_enum_framesizes(struct file *file, void *priv,
 					 struct v4l2_frmsizeenum *fsize)
 {
 	struct mxc_isi_video *video = video_drvdata(file);
+	const struct mxc_isi_pipe *pipe = video->pipe;
 	const struct mxc_isi_format_info *info;
 	unsigned int max_width;
 	unsigned int h_align;
@@ -1342,7 +1343,7 @@ static int mxc_isi_video_enum_framesizes(struct file *file, void *priv,
 	h_align = max_t(unsigned int, info->hsub, 1);
 	v_align = max_t(unsigned int, info->vsub, 1);
 
-	max_width = video->pipe->id == video->pipe->isi->pdata->num_channels - 1
+	max_width = (!pipe->bypass && pipe->id == pipe->isi->pdata->num_channels - 1)
 		  ? MXC_ISI_MAX_WIDTH_UNCHAINED
 		  : MXC_ISI_MAX_WIDTH_CHAINED;
 
